### PR TITLE
Adding a separate probe for the official speedtest-cli

### PR DESCRIPTION
**Pull request by flosoft**
I thought I'd give it a try to create a updated probe file that supports the official speedtest-cli.
This is by no means perfect, it still does upload and download for all transactions, but it's a starting point.

Tested it on my local smokeping.

https://github.com/mad-ady/smokeping-speedtest/issues/16

### CODE REVIEW

File: speedtestcli.pm [hunk 1]
```diff
@@ -135,34 +135,20 @@ sub pingone ($){
...
-            push @times, $normalizedvalue;
-            last;
-	    }
-	}
-	waitpid $pid,0;
-	close $errh;
-	close $inh;
-	close $outh;
+        my ($value) = /"$measurement":{"bandwidth":([0-9]*)/;
+        my $normalizedvalue = $value * 8;
```
Comment (mad-ady) [inline speedtestcli.pm:139]: Why are you multiplying by 8? Are you converting bytes to bits?

**Review by mad-ady — COMMENTED** [review]
I think it's good to have it independent of the speedtest-cli. I know there has been some talks about the naming between Ookla and speedtest-cli's developer. The probes should be independent.

File: speedtestcli.pm (hunk 1 above)
Comment (flosoft) [reply speedtestcli.pm:139]: Indeed. This way it aligns with the output from the prior probe. I checked this against the values on their results page.
The [following image](https://cloud.florianjensen.com/z8uxBjgY) is with me changing the probe from speedtest to "speedtestcli" on May 13, where data continues to align correctly while also matching the result on Speedtests.net for the same tests.

**Comment by flosoft** [comment]
Happy to follow any suggestion here for the naming. The other option I had in mind was `ookla-speedtest`.
What is interesting to see is that there is a huge performance difference when it comes to upload speeds measured, especially on servers that have a higher latency.

That said, we also see an improvement on the download side.

All probes were changed on the 13th of May:
A server close by (3ms):
[Download speeds](https://cloud.florianjensen.com/12uybnOw)
[Upload speeds](https://cloud.florianjensen.com/QwuKXmGK)

A server a bit further away (45-50ms)
[Download Speeds improvement](https://cloud.florianjensen.com/NQuDBPRn)
[Upload Speeds improvement](https://cloud.florianjensen.com/d5uv278v)

**Comment by mad-ady** [comment]
Wow, that's quite some difference... Can you confirm that you get about the
same values with their speedtest html5 client for those destinations?
I'll merge your changes and hope to test them myself.

Regarding upload/download being done in the same time - I had this issue
when I was running 3 smokeping instances on the same hardware and testing
speedtest via 3 different links - I would sometime get overlap between
tests that would interfere with eachother and I was able to "fix" it by
using the --download-only or --upload-only flags and have separate probes.
We'll see how it goes with this variant.

Thanks for your contribution!

**Comment by flosoft** [comment]
Yup, with the x8 factor it matches 1:1 to the results on the web.
You can confirm this by running smokeping in debug, where it'll give you the full JSON response with the web link.
It also matches the speeds I'm getting when running this manually on my machine.

In terms of the overlap, one thing I did was when starting the smokeping instance making sure that the random delay it sets is far apart between the upload and the download tests.

**Comment by mad-ady** [comment]
Could you share your configuration/random settings please?
Thanks for your contribution!

**Comment by flosoft** [comment]
Sample configuration:

Targets:
```
+ speedtest
menu = Speedtest.net
title = Speedtest.net probes

++ download_fr_gravelines_ovh
menu = Download OVH Gravelines
title = Download from OVH Cloud (Gravelines, France) [1050.10 km]
probe = speedtestcli-download
server = 25985
measurement = download
host = dummy.com

++ upload_fr_gravelines_ovh
menu = Upload OVH Gravelines
title = Upload from OVH Cloud (Gravelines, France) [1050.10 km]
probe = speedtestcli-upload
server = 25985
measurement = upload
host = dummy.com
```

Probes:
```
+ speedtestcli
binary = /usr/local/bin/ookla-speedtest
timeout = 300
step = 3600
offset = random
pings = 3
forks = 1

++ speedtestcli-download
measurement = download

++ speedtestcli-upload
measurement = upload
```

**Comment by mad-ady** [comment]
Hmm, this is odd - I was cleaning up the documentation and re-enabled the check for the binary (extracting the version), but the speedtest binary falls flat on its face when run with --version from inside smokeping:
```
DEBUG: terminate called after throwing an instance of 'std::logic_error'
Aborted (core dumped)
```
Stack overflow indicates trying to set a NULL string... Most likely it tries to read something missing from the environment.
I can extract it with this, though...
```
 strings /usr/bin/speedtest | egrep "^Speedtest by Ookla $" -B 1 | head -1
```
But it seems a convoluted way of doing things...

**Comment by mad-ady** [comment]
Ok, managed to find the problem. In my Docker environment speedtest would always segfault. It seems smokeping's environment didn't have a $HOME set and this caused the null pointer (the binary tries to read $HOME/.speedtest/). I added a fake $HOME in case it's missing in the probe to prevent the crash:  https://github.com/mad-ady/smokeping-speedtest/commit/aa78dcdbba76f2e81fc29f55c844e241488cb9eb

Steps to reproduce (also informed ookla via email):
```
 H frost  adrianp  ~  echo $HOME
/home/adrianp
 H frost  adrianp  ~  /usr/bin/speedtest --version
Speedtest by Ookla 1.0.0.2 (5ae238b) Linux/x86_64-linux-musl 5.3.0-51-generic x86_64

The official command line client for testing the speed and performance
of your internet connection.
 H frost  adrianp  ~  unset HOME
 H frost  adrianp  / | home | adrianp  echo $HOME

 H frost  adrianp  / | home | adrianp  /usr/bin/speedtest --version
terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_M_construct null not valid
Aborted (core dumped)
 H frost  adrianp  / | home | adrianp  134 
```

**Pull request merged**